presto-native-execution/presto_cpp/main/LinuxMemoryChecker.cpp

```diff
@@ -0,0 +1,151 @@
+  // However we noticed that this value was not closely related to the
+  // value in usage_in_bytes which is used to OOMKill. We then looked
+  // at all of the values in the stat file and decided that
+  // inactive_anon + active_anon moves closest to that of
+  // usage_in_bytes
+  //
+  // NOTE: We do not know if cgroup V2 memory.current is a fuzz
+  // value. It may be better than what we currently use. For
+  // consistency we will match cgroup V1 and change if
+  // necessary.
+  int64_t systemUsedMemoryBytes() override {
+    static std::string statFile;
+    static const boost::regex inactiveAnonRegex(
+        R"!(inactive_anon\s*(\d+)\s*)!");
+    static const boost::regex activeAnonRegex(R"!(active_anon\s*(\d+)\s*)!");
+    static const boost::regex memAvailableRegex(
+        R"!(MemAvailable:\s*(\d+)\s*kB)!");
+    static const boost::regex memTotalRegex(R"!(MemTotal:\s*(\d+)\s*kB)!");
+    static const char* cgroupV1Path = "/sys/fs/cgroup/memory/memory.stat";
+    static const char* cgroupV2Path = "/sys/fs/cgroup/memory.stat";
+
+    size_t memAvailable = 0;
+    size_t memTotal = 0;
+    size_t inactiveAnon = 0;
+    size_t activeAnon = 0;
+
+    // Find out what cgroup version (v1 or v2) we have based on the directory
+    // it's mounted.
+    if (statFile.empty()) {
+      struct stat buffer;
+      if ((stat(cgroupV1Path, &buffer) == 0)) {
+        statFile = cgroupV1Path;
+      } else if ((stat(cgroupV2Path, &buffer) == 0)) {
+        statFile = cgroupV2Path;
+      } else {
+        statFile = "None";
+      }
+      LOG(INFO) << fmt::format("Using memory stat file {}", statFile);
+    }
+
+    if (statFile != "None") {
+      folly::gen::byLine(statFile.c_str()) |
+          [&](const folly::StringPiece& line) -> void {
+        inactiveAnon =
+            extractNumericConfigValueWithRegex(line, inactiveAnonRegex);
+        activeAnon = extractNumericConfigValueWithRegex(line, activeAnonRegex);
+      };
+
+      // Unit is in bytes.
+      return inactiveAnon + activeAnon;
+    }
+
+    // Last resort use host machine info.
+    folly::gen::byLine("/proc/meminfo") |
+        [&](const folly::StringPiece& line) -> void {
+      memAvailable =
+          extractNumericConfigValueWithRegex(line, memAvailableRegex) * 1024;
+      memTotal = extractNumericConfigValueWithRegex(line, memTotalRegex) * 1024;
+    };
+    // Unit is in bytes.
+    return (memAvailable && memTotal) ? memTotal - memAvailable : 0;
+  }
+
+ private:
+  size_t extractNumericConfigValueWithRegex(
+      const folly::StringPiece& line,
+      const boost::regex& regex) {
+    boost::cmatch match;
+    if (boost::regex_match(line.data(), match, regex)) {
+      if (match.size() > 1) {
+        std::string numStr(match[1].str());
+        return std::stoul(numStr);
+      }
+    }
+    return 0;
+  }
+};
+
+namespace {
+static std::unique_ptr<PeriodicMemoryChecker> createLinuxMemoryChecker() {
```

Comment:
If this is only used once, we do not need a function. Let's inline the body of this function below.